### ISSUE TITLE
.ts file is being transpiled in different ways

### ISSUE THREAD

**Issue by ORESoftware** [Microsoft/TypeScript#24513]
**TypeScript Version:**  2.8.3

**Search Terms:** 
same file different results

**Code**

I have this source file:

```ts
import chai = require('chai');
import AssertStatic = Chai.AssertStatic;

export {InjectParam} from "./test-suite-params/inject/inject-param";
export {EachHookParam} from "./test-suite-params/each-hook/each-hook-param";
export {AllHookParam} from "./test-suite-params/all-hook/all-hook-param";
export {TestCaseParam} from "./test-suite-params/test-case/test-case-param";
export {ITestSuite} from 'suman-types/dts/test-suite';
export {ItFn, ITestDataObj} from 'suman-types/dts/it';
export {IDescribeFn, IDescribeOpts, TDescribeHook} from "suman-types/dts/describe";
export {IBeforeFn} from 'suman-types/dts/before';
export {IBeforeEachFn} from 'suman-types/dts/before-each';
export {IAfterFn} from 'suman-types/dts/after';
export {IAfterEachFn} from 'suman-types/dts/after-each';
export {DefineObject} from "./test-suite-helpers/define-options-classes";
export {DefineObjectContext as DefObjContext} from "./test-suite-helpers/define-options-classes";
export {DefineObjectTestCase as DefObjTestCase} from "./test-suite-helpers/define-options-classes";
export {DefineObjectAllHook as DefObjAllHook} from "./test-suite-helpers/define-options-classes";
export {DefineObjectEachHook as DefObjEachHook} from "./test-suite-helpers/define-options-classes";
```

Sometimes it transpiles to this (this is the desired output):

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var inject_param_1 = require("./test-suite-params/inject/inject-param");
exports.InjectParam = inject_param_1.InjectParam;
var each_hook_param_1 = require("./test-suite-params/each-hook/each-hook-param");
exports.EachHookParam = each_hook_param_1.EachHookParam;
var all_hook_param_1 = require("./test-suite-params/all-hook/all-hook-param");
exports.AllHookParam = all_hook_param_1.AllHookParam;
var test_case_param_1 = require("./test-suite-params/test-case/test-case-param");
exports.TestCaseParam = test_case_param_1.TestCaseParam;
var define_options_classes_1 = require("./test-suite-helpers/define-options-classes");
exports.DefineObject = define_options_classes_1.DefineObject;
var define_options_classes_2 = require("./test-suite-helpers/define-options-classes");
exports.DefObjContext = define_options_classes_2.DefineObjectContext;
var define_options_classes_3 = require("./test-suite-helpers/define-options-classes");
exports.DefObjTestCase = define_options_classes_3.DefineObjectTestCase;
var define_options_classes_4 = require("./test-suite-helpers/define-options-classes");
exports.DefObjAllHook = define_options_classes_4.DefineObjectAllHook;
var define_options_classes_5 = require("./test-suite-helpers/define-options-classes");
exports.DefObjEachHook = define_options_classes_5.DefineObjectEachHook;
```

_but sometimes it transpiles to this instead: ( this is the undesired output )_

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var inject_param_1 = require("./test-suite-params/inject/inject-param");
exports.InjectParam = inject_param_1.InjectParam;
var each_hook_param_1 = require("./test-suite-params/each-hook/each-hook-param");
exports.EachHookParam = each_hook_param_1.EachHookParam;
exports.AllHookParam = all_hook_param_1.AllHookParam;
var test_case_param_1 = require("./test-suite-params/test-case/test-case-param");
exports.TestCaseParam = test_case_param_1.TestCaseParam;
var test_suite_1 = require("suman-types/dts/test-suite");  // !!! this is causing a runtime error
exports.ITestSuite = test_suite_1.ITestSuite;
var it_1 = require("suman-types/dts/it");
exports.ItFn = it_1.ItFn;
exports.ITestDataObj = it_1.ITestDataObj;
var describe_1 = require("suman-types/dts/describe");
exports.IDescribeFn = describe_1.IDescribeFn;
exports.IDescribeOpts = describe_1.IDescribeOpts;
exports.TDescribeHook = describe_1.TDescribeHook;
var before_1 = require("suman-types/dts/before");
exports.IBeforeFn = before_1.IBeforeFn;
var before_each_1 = require("suman-types/dts/before-each");
exports.IBeforeEachFn = before_each_1.IBeforeEachFn;
var after_1 = require("suman-types/dts/after");
exports.IAfterFn = after_1.IAfterFn;
var after_each_1 = require("suman-types/dts/after-each");
exports.IAfterEachFn = after_each_1.IAfterEachFn;
var define_options_classes_1 = require("./test-suite-helpers/define-options-classes");
exports.DefineObject = define_options_classes_1.DefineObject;
var define_options_classes_2 = require("./test-suite-helpers/define-options-classes");
exports.DefObjContext = define_options_classes_2.DefineObjectContext;
var define_options_classes_3 = require("./test-suite-helpers/define-options-classes");
exports.DefObjTestCase = define_options_classes_3.DefineObjectTestCase;
var define_options_classes_4 = require("./test-suite-helpers/define-options-classes");
var define_options_classes_5 = require("./test-suite-helpers/define-options-classes");
exports.DefObjEachHook = define_options_classes_5.DefineObjectEachHook;
```

**Expected behavior:**
I would expect that given the same input file, and given the same version of tsc, the file to be transpiled to the same output. 

**Actual behavior:**

For some reason (probably a good one), the input is transpiled to different output.
The ultimate problem is that my program ends up trying to load .js files, but all there is to load is .d.ts files, so I get a 
**module not found error.**

Basically in the desired output, I don't have any `require('suman-types/dts/xxx')` calls, but in the undesired output, I have a few `require('suman-types/dts/xxx')` calls, and that is what is the problem.

**Comment by ORESoftware** [comment]
my only hypothesis at the moment, is that maybe in a certain scenario, `tsc` cannot load the `.d.ts` files, so it just assumes that they are .js files instead?

**Comment by mhegazy** [comment]
I am afraid this is pretty vague. please provide some concrete repro steps, configuration used, how you are building etc.. for us to be able to debug the issue.

**Comment by ORESoftware** [comment]
@mhegazy sure here is my tsconfig.json file:

```
{
  "compilerOptions": {
    "outDir": "dist",
    "skipLibCheck": true,
    "declaration": true,
    "baseUrl": ".",
    "typeRoots": [
      "./node_modules/@types",
      "./node_modules/suman-types/dts"
    ],
    "target": "es6",
    "module": "commonjs",
    "noImplicitAny": false,
    "removeComments": true,
    "allowUnreachableCode": true,
    "lib": [
      "es2015",
      "es2016",
      "es2017",
      "dom",
      "es5",
      "es2015.promise"
    ]
  },
  "compileOnSave": false,
  "include": [
    "src"
  ]
}
```

and I am just building with `tsc`

**Comment by ORESoftware** [comment]
It's possible that when it compiles, it is missing

```
"./node_modules/suman-types/dts"
```

**Comment by mhegazy** [comment]
the types should not change the generated output. so something else is going on. do you have a consistent repro?

**Comment by ORESoftware** [comment]
yeah I have a consistent repro, it's happening when my Github repo is running on Travis.

Do you want me to try to trim it down to just a few files?

**Comment by ORESoftware** [comment]
I think my original hypothesis is correct actually. I fixed it by including the 

```
   "./node_modules/suman-types/dts"
```

in my build. Does that make any sense?

**Comment by mhegazy** [comment]
> Do you want me to try to trim it down to just a few files?

that would be ideal. thanks!

**Comment by ORESoftware** [comment]
I know what's going on, it is failing if `"./node_modules/suman-types/dts"` is missing, but in some cases on CI/CD platforms, node_modules gets cached so it will succeed on second try.

I don't think we need a repro since I know what's happening, we just need to discuss why the transpile would be wrong if .d.ts files are missing.

**Comment by RyanCavanaugh** [comment]
> I don't think we need a repro

We do.

**Comment by ORESoftware** [comment]
I can get you the repro, but I wanted to discuss first if this is simply expected behavior - it might very well be. If .d.ts files are missing from the build, will `tsc` sometimes assume they are .js files or .ts files instead of .d.ts files?

**Comment by arciisine** [comment]
I'm not sure if this is related, but with 2.9.1 I'm running into an issue where previously working code, is now dropping a specific import causing the code to break.  I'm trying to narrow down to a better repro case.  The odd part, is the referenced variable from the import remains in the file, but the import clause does not.

**Comment by mhegazy** [comment]
If the export is type only, i.e. only an interface or a type alaise, and it is originating in a .d.ts file. the compiler needs to the .d.ts file to know whether to elide the export or not.

if this is the case you have, then that is expected. for anything else, no differences should be there.

**Comment by mhegazy** [comment]
@arciisine please file a new ticket and give us additional information about the issue you are running into.

**Comment by ORESoftware** [comment]
I created a REPRO for this:

https://github.com/ORESoftware/ts_bug_repro

the problem is real. whether the problem is a bug or a feature is TBD

**Comment by ORESoftware** [comment]
Basically in the readme file of the repro project I linked to above, I think `tsc` is doing the correct thing - if it can't find a .d.ts file, it then simply assumes that it is a missing .js or .ts file, so it generates the code to load that file. I think that if people care about this problem, they would simply listen to the `tsc` output for errors, and as long as `tsc` always complains that it cannot find a file, then it's the users fault for ignoring the `tsc` compilation error.

**Comment by ORESoftware** [comment]
As I mentioned the reason I am running `tsc` before running `npm install` is because of this chicken and egg problem:

https://stackoverflow.com/questions/50616253/catch-22-situation-with-tsc-and-npm-install

any insight into how to best solve that S.O. issue would good to know. Because that is what caused me to not run `npm install` in the first place.

**Comment by mhegazy** [comment]
thanks @ORESoftware for the repro. looks like the issue is what i have described  in https://github.com/Microsoft/TypeScript/issues/24513#issuecomment-393692390. for type only exports to be elided, the compiler needs to have access to the declaration site.
